### PR TITLE
Use celery signals to set request id automatically

### PR DESCRIPTION
**Pull request by bloodearnest**
Also, use task headers rather than thread local for celery enqueue metrics

### CODE REVIEW

File: talisker/celery.py [hunk 1]
```diff
@@ -79,48 +52,35 @@ def signal(sender, **kwargs):
...
+    task.talisker_timer.start()
+    id = task.request.talisker_request_id
+    if id is not None:
+        talisker.request_id.set(id)
 
 
 def task_postrun(sender, task_id, task, **kwargs):
-    task.__talisker_timer.stop()
+    task.talisker_timer.stop()
+    talisker.request_context.clear()
```
Comment (emgee) [inline talisker/celery.py]: Hmm, is there a talisker request_context to clear here? Where does the `postrun` signal execute - in the WSGI app or in the celery worker?

File: talisker/request_id.py [hunk 2]
```diff
@@ -53,8 +53,12 @@ def get():
 
 def set(id):
     """Sets id in both general request context, and specific logging dict"""
-    request_context.request_id = id
-    set_logging_context(request_id=id)
+    if id is None:
+        del request_context.request_id
+        request_context.extra.pop('request_id'), None
```
Comment (emgee) [inline talisker/request_id.py]: I'm guessing ", None" should be inside the func's parentheses? Suggests there's a test missing.

File: talisker/request_id.py [hunk 3]
```diff
@@ -53,8 +53,12 @@ def get():
 
 def set(id):
     """Sets id in both general request context, and specific logging dict"""
-    request_context.request_id = id
-    set_logging_context(request_id=id)
+    if id is None:
```
Comment (emgee) [inline talisker/request_id.py]: Setting to None and deleting feel like different operations but I'm fine if you're sure they're equivalent in the context of talisker.

File: talisker/celery.py [hunk 4]
```diff
@@ -79,48 +52,35 @@ def signal(sender, **kwargs):
...
+        talisker.statsd.get_client().timing(name, ms)
 
 
 def task_prerun(sender, task_id, task, **kwargs):
     name = 'celery.{}.run'.format(sender.name)
-    task.__talisker_timer = talisker.statsd.get_client().timer(name)
-    task.__talisker_timer.start()
+    task.talisker_timer = talisker.statsd.get_client().timer(name)
+    task.talisker_timer.start()
+    id = task.request.talisker_request_id
```
Comment (bloodearnest) [inline talisker/celery.py]: Yeah, this is celery dynamic programming madness. The task.request object proxies getattr to a dict, idk why </shrug>

File: talisker/celery.py (hunk 1 above)
Comment (bloodearnest) [inline talisker/celery.py]: It's an implicit context for the current request/job execution. It's either thread local storage, or greenlet local. Postrun runs in the celery worker, straight after the task completes.

Basically, the only way to provide request_id (and other stuff) to all logging calls is by sticking it in an implicit context. Sucks, but talisker's been that way from the start, as stdlib logging with it's globals is used everywhere.

File: talisker/request_id.py (hunk 3 above)
Comment (bloodearnest) [inline talisker/request_id.py]: Well, if we set to None, we'll get "request_id=None" on every log line, so...

File: talisker/request_id.py (hunk 2 above)
Comment (bloodearnest) [inline talisker/request_id.py]: Huh, yes, good catch

**Pull request merged**